### DISCUSSION TITLE
scanner status on WinRT backend

### DISCUSSION THREAD

**Discussion by aliencaocao** [hbldh/bleak#638] [Q&A] [answered]
Hi all, is there a replacement for the status attribute of the new WinRT backend? I am following the official doc but that was only for dotnet backend, and now when I try to access the status attribute of the scanner, it give me `'BleakScannerWinRT' object has no attribute 'status'`.

Basically, I need to stop the scanning process and the scanner the moment it found **any** device or encountered any error (to be more exact, get the code to continue executing after the `loop.run_until_complete(run())` since it is blocking if the run coroutine does not end) and at the same time give it a timeout of 5seconds so that it stops after 5seconds of finding no device at all. My current approach is using a while loop that checks the status of the scanner, which is no longer possible after WinRT is implemented. Any suggestion on how do I do this now?

What I have currently (which is not working):

```
        async def run():
            async def early_stopping(_, _1):
                await scanner.stop()
                self.passed = True
            scanner = BleakScanner()
            scanner.register_detection_callback(early_stopping)
            await scanner.start()
            while scanner.status not in [3, 4]:  # no longer working in WinRT
                pass
        import asyncio
        loop = asyncio.get_event_loop()
        try:
            loop.run_until_complete(run())
        except Exception as e:
            logger.error(f'Bluetooth scanning test failed due to error: {e}')
            self.passed = False
```

**Comment by dlech** [comment, marked as answer]
You could try https://github.com/hbldh/bleak/blob/develop/examples/scanner_byname.py. You can replace the lambda filter function with something that makes sense for your device if matching by name is not suitable.

**Comment by aliencaocao** [comment]
Maybe I didn't phrase it properly, but what I meant was the moment it found **any** device, it stops. Not a particular one and I have no requirements for whatever it finds.

**Reply by dlech** [reply]
in that case `lambda d, ad: True` will match all devices.

**Reply by aliencaocao** [reply]
I actually managed to access the status through scanner.watcher.status, however the scanner.register_detection_callback is not working. The scanner just permanently continues until I disable the Bluetooth hardware on my OS, then it will exit with an error and print out a list of found devices (so it actually found devices but not triggering the callback function at all.

**Reply by aliencaocao** [reply]
As for your suggestion, it gave me error: 'NoneType' object has no attribute 'stop'

**Reply by dlech** [reply]
It would be best to not access the OS-specific internals of Bleak. The code you gave never calls `scanner.stop()` so seems expected that scanning would never stop. The example I linked takes care of this for you automatically with much less code that you have to write yourself.

**Reply by aliencaocao** [reply]
The code i gave have a callback function that includes `await scanner.stop()`, but that callback fails to trigger at all. I switched it to simply printing device rssi per documentation's example and nothing would be printed.
The reason why I have to access the status is I need a while loop to act as a blocking code for the program to wait till it found any device, else it would just continue to execute no matter what.

**Reply by aliencaocao** [reply]
I am not sure how do I implement something like a timeout with your example. For the while loop method I used, I could have a timer started and inside the loop checks for the time delta, when it reaches the timeout, it will simply break the loop

**Reply by aliencaocao** [reply]
For the callback issue: I figured out that it was due to my while loop being inside the async run() function. The moment I remove it, the callback works. However I am still not sure why would it not work if I have the while loop there. I already called await scanner.start() before the loop so it should already be scanning.

**Comment by dlech** [comment]
`find_device_by_filter()` has a timeout argument.

docs: https://bleak.readthedocs.io/en/latest/api.html#bleak.backends.scanner.BaseBleakScanner.find_device_by_filter

**Reply by aliencaocao** [reply]
Ahhh sorry silly me did not check the doc for that function...
I just did a test run and it works as I want it to be, thank you for the prompt replies and sorry for taking up for time!
